src/wms/src/main/java/org/geoserver/wms/NearestMatchFinder.java

```diff
@@ -0,0 +1,469 @@
+import org.geotools.util.Range;
+import org.opengis.coverage.grid.GridCoverageReader;
+import org.opengis.filter.Filter;
+import org.opengis.filter.FilterFactory;
+import org.opengis.filter.expression.Literal;
+import org.opengis.filter.expression.PropertyName;
+
+import java.io.IOException;
+import java.text.ParseException;
+import java.util.Date;
+import java.util.TreeSet;
+
+import static org.geoserver.wms.NearestMatchFinder.FilterDirection.HIGHEST_AMONG_LOWERS;
+import static org.geoserver.wms.NearestMatchFinder.FilterDirection.LOWEST_AMONG_HIGHER;
+
+/**
+ * Support class to find the nearest match to a given dimension value
+ */
+public abstract class NearestMatchFinder {
+
+    /**
+     * Made available only for testing purposes, as there is no simple (non structured) reader supporting time
+     * but better to work off the "reference" one (the image mosaic)
+     */
+    public static boolean ENABLE_STRUCTURED_READER_SUPPORT = true;
+
+    /**
+     * Returns an implementation of {@link NearestMatchFinder} optimized for the given resource type, or throws
+     * an {@link IllegalArgumentException} in case the resource type is not supported
+     *
+     * @param info
+     * @param dimensionInfo
+     * @param dimensionName
+     * @return
+     * @throws IOException
+     */
+    public static NearestMatchFinder get(ResourceInfo info, DimensionInfo dimensionInfo, String dimensionName) throws
+            IOException {
+        Class dataType = getDataTypeFromDimension(info, dimensionName);
+        try {
+            AcceptableRange acceptableRange = AcceptableRange.getAcceptableRange(dimensionInfo.getAcceptableInterval
+                    (), dataType);
+            if (info instanceof FeatureTypeInfo) {
+                FeatureTypeInfo featureType = (FeatureTypeInfo) info;
+                return new Vector(featureType, dimensionInfo.getAttribute(), dimensionInfo.getEndAttribute(),
+                        acceptableRange, dataType);
+            } else if (info instanceof CoverageInfo) {
+                GridCoverageReader reader = ((CoverageInfo) info).getGridCoverageReader(null, null);
+                if (reader instanceof StructuredGridCoverage2DReader && ENABLE_STRUCTURED_READER_SUPPORT) {
+                    StructuredGridCoverage2DReader structured = (StructuredGridCoverage2DReader) reader;
+                    DimensionDescriptor dd = getDimensionDescriptor(structured, dimensionName);
+                    return new StructuredReader(structured, dd.getStartAttribute(), dd.getEndAttribute(), 
+                            acceptableRange, dataType);
+                } else if (reader instanceof GridCoverage2DReader) {
+                    return new Reader((GridCoverage2DReader) reader, acceptableRange, dimensionName, dataType);
+                }
+            }
+        } catch (ParseException e) {
+            throw new ServiceException("Failed to apply nearest match search", e);
```

Comment:
Adding the ``info`` to the message may help contextualize.

Reply:
You mean the layer name?